resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java

```diff
@@ -888,15 +889,64 @@ protected void restoreMarkers(IResource resource, boolean generateDeltas, IProgr
 			}
 			return;
 		}
-		IProject[] projects = ((IWorkspaceRoot) resource).getProjects(IContainer.INCLUDE_HIDDEN);
-		for (IProject project : projects)
-			if (project.isAccessible())
-				markerManager.restore(project, generateDeltas, monitor);
+		forEachProjectInParallel(monitor, project -> {
+			if (project.isAccessible()) {
+				markerManager.restore(project, generateDeltas, null);
+			}
+		});
 		if (Policy.DEBUG_RESTORE_MARKERS) {
 			Policy.debug("Restore Markers for workspace: " + (System.currentTimeMillis() - start) + "ms"); //$NON-NLS-1$ //$NON-NLS-2$
 		}
 	}
 
+	@FunctionalInterface
+	public interface CoreConsumer<T> {
+		/**
+		 * Performs this operation on the given argument.
+		 *
+		 * @param t the input argument
+		 */
+		void accept(T t) throws CoreException;
+	}
+
+	private void forEachProjectInParallel(IProgressMonitor m, CoreConsumer<IProject> consumer) throws CoreException {
+		IProject[] projects = workspace.getRoot().getProjects(IContainer.INCLUDE_HIDDEN);
+		if (projects.length == 0) {
+			return;
+		}
+		SubMonitor subMointor = SubMonitor.convert(m, projects.length);
+		IStatus[] stats;
+		// Never use a shared ForkJoinPool.commonPool() as it may be busy with other tasks, which might deadlock.
+		// Also use a custom ForkJoinWorkerThreadFactory, to prevent issues with a
+		// potential SecurityManager, since the threads created by it get no permissions.
+		// See https://github.com/eclipse-platform/eclipse.platform/issues/294
+		ExecutorService executor = new ForkJoinPool(ForkJoinPool.getCommonPoolParallelism(),
+				pool -> new ForkJoinWorkerThread(pool) {
+					// anonymous subclass to access protected constructor
+				}, null, false);
+		try {
+			stats = executor.submit(() -> Arrays.stream(projects).parallel().map(project -> {
+				subMointor.split(1);
+				try {
+					consumer.accept(project);
+				} catch (CoreException e) {
+					return new ResourceStatus(IResourceStatus.FAILED_READ_METADATA, project.getFullPath(),
+							"Error with project " + project.getName(), e); //$NON-NLS-1$
+				}
+				return null;
+			}).filter(Objects::nonNull).toArray(IStatus[]::new)).get();
```

Comment:
You must not call `get()` here as it blocks, and you don't want to be blocked. Instead you should either return a `Stream<..>` with an `onClose( ... )`, a `Future<..>` or similar or even better a `CompletableFuture<..>` then you can use that in an event driven non blocking way.